### ISSUE TITLE
Vulkan backend on desktop platforms

### ISSUE THREAD

**Issue by skyline75489** [pytorch/pytorch#55879]
Currently Vulkan backend is only available on Android. Is there anything specific stopping us from porting Vulkan to desktop platforms (Windows/Linux/macOS) ?

**Comment by jbschlosser** [comment]
I believe Vulkan support on desktop platforms already exists at some level: https://pytorch.org/tutorials/prototype/vulkan_workflow.html

**Comment by skyline75489** [comment]
@jbschlosser Thanks for the information! The Vulkan build does not seem to work on Windows, though. I'll try to see if I can make it work on Windows. Is there a significant feature gap that prevent us from recommending people to use it? Or it's just CUDA is good enough?

**Comment by ahirner** [comment]
Hello @skyline75489 , any progress with your build attempts? 
I linked a specific issue that I encountered on master.

**Comment by skyline75489** [comment]
@ahirner I had a [working branch](https://github.com/skyline75489/pytorch/tree/chesterliu/dev%2Fvulkan-on-win) but unfortunely it's bit stale. If you look into it I think you'll find something useful.

**Comment by Nine-H** [comment]
Is there a roadmap to get vulkan onto desktop even if it's just in nightly? I have an AMD gpu on linux and ROCm is just grot: mandating specific kernels, specific distros, and having strange compatibility issues with mesa's open gallium driver.

**Comment by Matthew-Bradley** [comment]
I'd love to see this. Right now there's CUDA on nvidia and ROCm on (some of) AMD......but every GPU architecture under the sun is getting a vulkan driver.

Mobile GPUs for smartphones/chromebooks/laptops, but also intel GPUs and AMD GPUs (not all of which are supported by ROCm). Eventually there'll also be one for Linux on apple silicon, and for whatever LibreSoC produce.

Vulkan is quickly becoming the lingua-franca of the GPU world so supporting it means being able to run pytorch in lots of new places.

**Comment by orowith2os** [comment]
There was some work done over at https://github.com/pytorch/pytorch/pull/61431, but it was closed and I'm not sure what the current status of this is.

If Vulkan works on Android, will it also work on desktop Linux easily? Or are the two too different for that?

I'm not interested in Windows or macOS support here, as my main concern is Linux, but having some status update for those two would be appreciated as well.

**Comment by danielzgtg** [comment]
Yes, it should work on desktop Linux easily. Android-x86/anbox/Waydroid can sit alongside your desktop environment, and this shows how it's mostly the same when a container/VM separates them. Vulkan might not be implemented yet, but should be soon. Since you should be able to run PyTorch Vulkan through Android on desktop Linux, it is reasonable to conclude that PyTorch Vulkan should be able to run on desktop Linux directly.

The main differences I see are the libc and the graphics. I doubt the absence of the Android Java part, or the absence of the Android security restrictions will make a difference. Android and Linux are both x86 and ARM, so if that difference is a problem, a separate issue should be created for x86 Android. Android is bionic not glibc, and I see frequent breakages even between different versions of glibc. Android has some Android-specific GPU APIs like RenderScript but I don't think those are used as they are deprecated, and even if they are, they can be ported over as it's all Linux. SurfaceFlinger with a proprietary driver stack is used on Android instead of X11/Wayland with Mesa. Unless PyTorch's Vulkan implementation is somehow coupled to graphics for some weird reason, Vulkan should be relatively headless and independent.

I used to be excited for the Android Vulkan feature until I saw the problems the OpenCL implementation encountered. Basically only a few image-based tasks worked. AMD ROCm's CUDA has most things working and the setup is still painfully difficult. The greatest challenge would be the monumental undertaking in achieving feature parity with the massive amount of CUDA kernels. Then there would be the problem of third party libraries assuming GPU is always cuda so that this Vulkan implementation should follow ROCm in pretending to be CUDA.

**Comment by Matthew-Bradley** [comment]
Since my last comment the OpenCL landscape has changed drastically, with the introduction of rusticl. It provides a high-quality OpenCL 3.0 implementation, soon to be on every GPU platform including on top of existing vulkan drivers.

**Comment by redthing1** [comment]
Anyone knows of an update for this? Is Vulkan on desktop working well?

**Comment by dproworld** [comment]
Vulkan is superior, allows cross-vendor multi gpu vram sharing. What's stopping this?? even llama.cpp has had it for forever now!
